pex/pex_bootstrapper.py

```diff
@@ -18,122 +18,159 @@
+    #   https://github.com/pantsbuild/pex/issues/430
+    return min(candidate_interpreters)
 
-def maybe_reexec_pex(compatibility_constraints):
-  """
-  Handle environment overrides for the Python interpreter to use when executing this pex.
+
+def maybe_reexec_pex(compatibility_constraints=None):
+  """Handle environment overrides for the Python interpreter to use when executing this pex.
 
   This function supports interpreter filtering based on interpreter constraints stored in PEX-INFO
-  metadata. If PEX_PYTHON is set in a pexrc, it attempts to obtain the binary location of the
-  interpreter specified by PEX_PYTHON. If PEX_PYTHON_PATH is set, it attempts to search the path for
-  a matching interpreter in accordance with the interpreter constraints. If both variables are
-  present in a pexrc, this function gives precedence to PEX_PYTHON_PATH and errors out if no
-  compatible interpreters can be found on said path.
+  metadata. If PEX_PYTHON is set it attempts to obtain the binary location of the interpreter
+  specified by PEX_PYTHON. If PEX_PYTHON_PATH is set, it attempts to search the path for a matching
+  interpreter in accordance with the interpreter constraints. If both variables are present, this
+  function gives precedence to PEX_PYTHON_PATH and errors out if no compatible interpreters can be
+  found on said path.
 
   If neither variable is set, we fall back to plain PEX execution using PATH searching or the
   currently executing interpreter. If compatibility constraints are used, we match those constraints
   against these interpreters.
 
-  :param compatibility_constraints: list of requirements-style strings that constrain the
-  Python interpreter to re-exec this pex with.
+  :param compatibility_constraints: optional list of requirements-style strings that constrain the
+                                    Python interpreter to re-exec this pex with.
   """
-  if os.environ.pop('SHOULD_EXIT_BOOTSTRAP_REEXEC', None):
+
+  current_interpreter = PythonInterpreter.get()
+
+  # NB: Used only for tests.
+  if '_PEX_EXEC_CHAIN' in os.environ:
+    flag_or_chain = os.environ.pop('_PEX_EXEC_CHAIN')
+    pex_exec_chain = [] if flag_or_chain == '1' else flag_or_chain.split(os.pathsep)
+    pex_exec_chain.append(current_interpreter.binary)
+    os.environ['_PEX_EXEC_CHAIN'] = os.pathsep.join(pex_exec_chain)
+
+  current_interpreter_blessed_env_var = '_PEX_SHOULD_EXIT_BOOTSTRAP_REEXEC'
+  if os.environ.pop(current_interpreter_blessed_env_var, None):
     # We've already been here and selected an interpreter. Continue to execution.
     return
 
-  target = None
-  with TRACER.timed('Selecting runtime interpreter based on pexrc', V=3):
+  with TRACER.timed('Selecting runtime interpreter', V=3):
     if ENV.PEX_PYTHON and not ENV.PEX_PYTHON_PATH:
       # preserve PEX_PYTHON re-exec for backwards compatibility
       # TODO: Kill this off completely in favor of PEX_PYTHON_PATH
```

Comment:
This something that should probably happen soon. Not sure if we have a deprecation policy around this but it is redundant with PPP and  more complexity to support than necessary.

Reply:
I think the bit needed to kick off deprecation / removal is expanding "it's redundant" in an issue. FWICT it is not simply redundant. Today I can `PEX_PYTHON=python2 ...` to ensure a pex runs under some python2 interpreter on any machine. I can't do that with PEX_PYTHON_PATH since its entries are full paths or path prefixes, both of which are machine specific.